Sources/ClientRuntime/Interceptor/Interceptors.swift

```diff
@@ -0,0 +1,448 @@
+        self.interceptors.append(AnyInterceptor(modifyBeforeDeserialization: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadBeforeDeserialization(
+        _ interceptorFn: @escaping (any BeforeDeserialization<InputType, RequestType, ResponseType, AttributesType>)
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readBeforeDeserialization: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterDeserialization(
+        _ interceptorFn: @escaping (
+            any AfterDeserialization<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        )
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterDeserialization: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addModifyBeforeAttemptCompletion(
+        _ interceptorFn: @escaping (
+            any MutableOutputAfterAttempt<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        )
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(modifyBeforeAttemptCompletion: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterAttempt(
+        _ interceptorFn: @escaping (any AfterAttempt<InputType, OutputType, RequestType, ResponseType, AttributesType>)
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterAttempt: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addModifyBeforeCompletion(
+        _ interceptorFn: @escaping (
+            any MutableOutputFinalization<InputType, OutputType, RequestType, ResponseType, AttributesType>
+        ) async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(modifyBeforeCompletion: interceptorFn))
+    }
+
+    /// - Parameter interceptorFn: The closure to use as the Interceptor hook.
+    public mutating func addReadAfterExecution(
+        _ interceptorFn: @escaping (any Finalization<InputType, OutputType, RequestType, ResponseType, AttributesType>)
+            async throws -> Void
+    ) {
+        self.interceptors.append(AnyInterceptor(readAfterExecution: interceptorFn))
+    }
+}
+
+/// This extension is for the convenience of Orchestrator to call the same hook on all interceptors.
+/// Note that it doesn't conform to Interceptor because it is backed by AnyInterceptor, which also
+/// doesn't conform to Interceptor.
+///
+/// These methods throw the last error that occurred when executing a given hook, so if multiple
+/// interceptors fail, only the last error is thrown. The rest are logged (TODO)
```

Comment:
We have a goal to remove all TODOs in our SDK, if we plan to log the additional errors let's either include that logic or create a tracking ticket for that work and pull it into sprint

Reply:
Nice catch, forgot about this one.

Reply:
Done